Added ratechange cooldown to fix #72 #521 and #584

ChadBailey opened the pull request:
Based on user feedback and my own testing, I'm confident this will fix #584 which has been reported by many users. This should also fix #72 and fix #521 (NOTE: Originally #523 was claimed to be fixed by this in error, that one appears to be related to the apple.tv shadowroot bug)

~I realize in the past we [went back and forth](https://github.com/igrigorik/videospeed/pull/547#issuecomment-569454746) on implementing this or not, however, with one of the largest video sites on the Internet now being impacted I think it is time to bring this fix in. Implementing it with the warnings clearly shown seems to be the lesser of the two evils.~

~To re-iterate, the only known side effect of this fix is that adjusting using the website's controls (assuming it has any) will not update the controller's speed. Adjusting using VSC will immediately fix this by setting the speed to what the VSC says. This is unavoidable, since all ratechange listeners are blocked including ours.~

Edit: See below for notes on the new technique. The caveats listed above no longer apply.

igrigorik commented on inject.js:
```diff
@@ -18,7 +18,12 @@ var tc = {
       teams.microsoft.com
-      `.replace(regStrip, "")
+    `.replace(regStrip, ""),
+    blacklistrc: `\
```
nitpick: let's use a descriptive name. it's ok to be a bit more verbose if it helps clarity in the code.

blacklistRateChange or something like that.

ChadBailey replied:
I'm usually the one saying this, so I didn't follow my own advice lol. I'll update it.

ChadBailey replied:
This has been removed entirely with the latest version

igrigorik commented on options.js:
```diff
@@ -23,6 +23,11 @@ var tcDefaults = {
     teams.microsoft.com
+  `.replace(regStrip, ""),
+  blacklistrc: `\
+    twitch.tv
```
Are we sure about this? It seems like they have a good reason to increase rate (to catch up the stream) and we'd break that default behavior. It's not a bug on our part that the video speed jumps to 1.03.. it's exactly what twitch player intends.

ChadBailey replied:
I thought I was sure, but now I'm not so sure anymore...

In my testing I never experienced the playing changing to 1.03x, I only experienced it bouncing back to 1.0x after attempting to change like shown in the gif in #584

ChadBailey replied:
After fairly extensive testing with this fix disabled, I never experienced twitch.tv adjusting the playback rate of the video.

The only time I experienced an adjustment was when the playback buffer went beyond 30seconds, it will occasionally skip to a more recent time.

![image](https://user-images.githubusercontent.com/15102816/75684558-a0a3ec00-5c5e-11ea-9323-809ff56bf2e2.png)

What I suspect may be going on is some kind of adverse effect of using VSC without the fix applied. The original poster did explain the speed would never be fixed from 1.03 causing him to reload the entire page to fix. I would expect if this was intended behavior, it would auto-resolve once the playback has caught back up with the buffer.

ChadBailey replied:
This is odd, it happened quickly when I tested again using the normal production version of the addon found in the Chrome store... and it did adjust back as you mentioned, so it felt like intentional behavior.

![image](https://user-images.githubusercontent.com/15102816/75686272-9800e500-5c61-11ea-9b62-98e0a6018674.png)

I'm now left wondering if something else in the updates was impacting this, or if I just got unlucky with my first tests. I'll keep poking at it and see what I can find out.

ChadBailey replied:
Sorry for my monologue, found this one interesting. Either I messed up my tests the first go-around, or what I think is happening is this feature appears to be present on some streams but not others.

Visiting the same stream using the test environment, I was able to immediately see the speed change to 1.03 as it had in the production version of VSC. I didn't make any code or configuration changes between these tests. I suspect we may have stumbled upon some AB testing twitch is doing to see if this feature is worth/ready to push out to the entire user base.

![image](https://user-images.githubusercontent.com/15102816/75687308-39d50180-5c63-11ea-82ee-8bb1c9108874.png)

Knowing what I know now, I have a few thoughts.

First, we are already interfering with the default twitch playback rate - which may have been why they in turn be why they created a more aggressive mechanism for adjusting speed.

This is done on [line 149 of inject.js](https://github.com/igrigorik/videospeed/blob/master/inject.js#L149)

This has been a part of the behavior [since the initial commit](https://github.com/igrigorik/videospeed/blame/f1827ba63060255ba119431a00d6a26964058f84/inject.js#L61)

Second thought, because the fix does not actually block a ratechange from happening, this fix does not prevent the default behavior of twitch.tv. I validated this by adding a console.log statement before blocking the propagation

```
  if (isRateChangeBlocked()) {
    document.body.addEventListener(
      "ratechange",
      function(event) {
        console.log(
          "Speed changed to: " +
            event.target.playbackRate +
            " (event propogation blocked)"
        );
        event.stopImmediatePropagation();
      },
      true
    );
  }
```

With all of that known, I'm not sure which direction to go. I two valid paths we could take
1. Add twitch.tv to the blacklist as a default, and users are able to move it to the ratechange blocklist if they so desire
2. Add twitch.tv to the ratechange blacklist as a default, and users will be able to use the extension as they expect, without a visual indication of the speed changing to 1.03 then back to 1.0x once the stream is at 1.5 seconds behind the broadcaster.

I suppose the first thing we need to figure out is if we would like to disable the behavior of manually setting the speed to 1.0x every time we initialize videos - which is totally unrelated to the proposed changes here.

igrigorik replied:
> Add twitch.tv to the ratechange blacklist as a default, and users will be able to use the extension as they expect, without a visual indication of the speed changing to 1.03 then back to 1.0x once the stream is at 1.5 seconds behind the broadcaster.

If we implement ratechange blacklist and add twitch by default, my expectation is that the above behavior would not happen: the global listener would catch and block any updates initiated by non VSC logic. Is that not so?

ChadBailey replied:
Good news! I came up with a new method that nullifies the need for a blacklist and should work on any website without interfering!

I think this is a much better fix that addresses your concerns. The new logic works on a timer. When you explicitly change the speed using video speed controller, it will block on rate change events that aren't generated by VSC for a 1 second window. That 1 second window is then refreshed every time a speed adjustment is done in VSC. Once you have stopped adjusting the speed for 1 second, it resumes normal operation.

In my tests, this fully blocks the "bounce back" effect with no ill effect on the regular operation of websites and the fix is seamless.

ChadBailey replied:
Your expectation is actually not correct because the rate change command does not originate from a rate change listener.

I have come up with an even better solution that nullifies the need for a rate change block list. See my comments below.

ChadBailey commented:
Good news! I came up with a new method that nullifies the need for a blacklist and should work on any website without interfering! I think this is a much better fix that addresses your concerns.

In the new logic, if you make no attempt to adjust the speed using VSC, everything continues working the same way it does today. Once you issue a command to adjust the speed using VSC, this begins a 1-second timer. Rate change events will now be blocked until the 1-second timer completes. That 1-second timer is then refreshed every time a speed adjustment is requested using VSC. Once you have stopped adjusting the speed in VSC for 1 full second, it resumes normal operation.

In my tests, this fully blocks the "bounce back" effect with no ill effect on the regular operation of websites and the fix is seamless.

igrigorik commented on inject.js:
```diff
@@ -551,6 +578,14 @@ function initializeNow(document) {
+  var speedvalue = speed.toFixed(2);
+  video.playbackRate = Number(speedvalue);
+  var speedIndicator = controller.shadowRoot.querySelector("span");
+  speedIndicator.textContent = speedvalue;
```
Do we still need these two lines given updated logic?

ChadBailey replied:
Thanks for pointing this out...

Yes, actually this is even more essential now as the ratechange event is blocked, however, this made me realize that now the logic inside of our ratechange event would never be triggered by VSC which I think is the -exact opposite- of the desired behavior.

I have moved this into its own function and now it's triggered irrespective of if the speed change request comes from VSC or if it comes from the site itself. This does bring in to question whether we should be storing lastspeed when it's the website that requested the change rather than VSC, but I've preserved the existing behavior.

If this needs to be changed to only store speeds originating from requests of VSC I certainly don't mind making that change but think it may be better to split into its own bug report since this is how it behaves today.

ChadBailey commented:
I want to make sure we are on the same page as far as what this change actually does so I've prepared a gherkin formatted feature here. I hope this helps clarify exactly how this (should) behave

```feature
Feature: Bounce Back Protection

    As a user of the video speed controller extension, I want to be able to
    adjust the video playback speed on websites without it being changed back
    immediately so that I can set the speed I prefer

    Scenario: Website sets playback rate on ratechange event
      Given a website twitch.tv that sets the playback rate on change
       When the speed is changed using video speed controller
       Then the ratechange event is triggered
        And the event is blocked
        And the playback rate does not change again

    Scenario: Website explicitly sets playback rate
      Given a website twitch.tv that sets the playback rate on change
        And the user has not used video speed controller to adjust the speed within the last 1000ms
       When the speed is changed and the origin is not video speed controller
       Then the ratechange event is triggered
        And the event is not blocked

    Scenario: Website does not set playback rate on ratechange event
      Given a website youtube.com that does not set the playback rate on change
       When the speed is changed using video speed controller
       Then the ratechange event is triggered
        And the event is blocked
```

ChadBailey commented:
Don't merge this yet, I'm still testing and it seems to be undoing the fix for youtube changing the playback rate to 1 when you pause/play. I'm guessing/hoping this is an easy fix, but I haven't had time to look closer yet

ChadBailey commented:
I just spent an inordinate amount of time trying to figure out why YouTube will remember the last playback rate irrespective of the "Remember Playback Speed" setting... only to find out that's YouTube's default behavior lol.

I think this is now ready to be merged, but do feel free to test it out for yourself

jedawson commented on manifest.json:
```diff
@@ -1,7 +1,7 @@
   "name": "Video Speed Controller",
   "short_name": "videospeed",
-  "version": "0.6.0",
+  "version": "0.6.1",
```
did you do this:
```bash
git tag -a v0.6.1 -m "bump to v0.6.1"
git push origin v0.6.1
```

ChadBailey replied:
I did not... I'll do that

ChadBailey replied:
Tagged, thanks for the tip

igrigorik commented on inject.js:
```diff
@@ -132,56 +163,54 @@ function defineVideoController() {
-
-    target.addEventListener(
-      "ratechange",
-      (this.handleRatechange = function(event) {
```
If we're removing the listener it means our controller will no longer pickup ratechange events from other sources.. e.g. if someone uses built-in YT speed controls, the displayed speed will be out of sync. 

That seems like a regression. Thoughts>

ChadBailey replied:
I actually didn't remove it, I just combined the two. The existing ratechange listener was placed in the DOM at the location of the video element (this seems wrong to say, but based on my observations this seems to be how an event propagates through the DOM - beginning at the root/document level then flowing down). This was too late, so I had to add another one ahead of it.

I now have combined them, so even if downstream ratechange listeners happen to be blocked by the edge case the controller is updated with the new speed. I will update the gherkin feature to explain the new behavior (the only change is very slight in the edge case)

ChadBailey commented on inject.js:
```diff
@@ -334,9 +359,38 @@ function initializeWhenReady(document) {
+    "ratechange",
+    function(event) {
+      controller = event.target.parentElement.querySelector(".vsc-controller");
+      rateChanged(controller);
```
I moved the original logic found in our ratechange into this function `rateChanged`. Since it is no longer being executed as part of controller's object model, I had to refactor it to use the event to find the controller rather than simply being able to use this.video. This allowed me to combine the two.

Originally, the `rateChanged` function was critical because it was manually triggered by speed adjustments via `setSpeed`, however, after combining them this is no longer a critical step.

ChadBailey commented:
After much testing, I noticed I was getting intermittent errors in the error log pointing to trying to initialize before the document was ready.

I've moved this a little further back in hopes that it will strike a good balance, but since this is by its very nature a race condition I can't be sure this won't penalize us in some situations. It does seem to be working fine on YouTube and Twitch.tv, but I'd like to give it a little more testing time to be sure I don't notice any ill effects. Once I feel confident I'll report back and request a review.

Sorry for so much back and forth, this has been a bit of a complex problem to solve.

ChadBailey commented:
I have not experienced any ill effects from these changes in my usage. Ready for review

ChadBailey commented on inject.js:
```diff
@@ -626,31 +645,9 @@ function setSpeed(controller, video, speed) {
   var speedvalue = speed.toFixed(2);
   video.playbackRate = Number(speedvalue);
   refreshCoolDown();
-  rateChanged(controller);
```
This call is no longer needed now that the ratechange blocker (to stop "bouncing" playback rate) and ratechange listener (for updating the controller's speed value) have been combined into one. setSpeed inherently calls the logic in rateChanged by virtue of the listener.

ChadBailey commented:
All issues have been addressed and I've been using this for a while without ill effect. Any objections to merging?

igrigorik commented:
👍

On Sat, Mar 21, 2020, 4:44 PM Chad Bailey <notifications@github.com> wrote:

> All issues have been addressed and I've been using this for a while
> without ill effect. Any objections to merging?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/igrigorik/videospeed/pull/604#issuecomment-602120414>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAACTHAHVO3WC7FGFQUHIP3RIVGPFANCNFSM4K7LYIQA>
> .
>
